### PR TITLE
Load rule ignores from external text file

### PR DESCRIPTION
**Pull request by ssbarnea**
Ansible-lint will load skip rules from `.ansible-lint-ignore` file that is adjacent to its config file. The file format is very simple, containing the filename and the rule to be ignored. It also supports comments starting with `#`.

- Adds ability to load ignores from file
- Implements `--generate-ignore` for dumping ignores for current violations
- Adds documentation about ignores
- Change summary message to point to documentation for ignores
- This feature also deprecates progressive mode due to its performance and testing problems, like testing of pull-requests with multiple commits.

Fixes: #1584
Fixes: #3012
Closes: #2934
Related: https://github.com/ansible/ansible-lint/issues/2880

### CODE REVIEW

**Comment by apatard** [comment]
Typo in the commit message title ? :)

File: .ansible-lint [hunk 1]
```diff
@@ -44,7 +44,10 @@ use_default_rules: true
 # rulesdir:
 #   - ./rule/directory/
 
-# Ansible-lint completely ignores rules or tags listed below
+# Ansible-lint is able to recognize and load skip rules stored inside
+# `tests/sanity/ignore.txt` files (same as ansible-test). To skip
```
Comment (mattclay) [inline .ansible-lint:48]: Since this is unrelated to the `ansible-test` `sanity` test, I don't think it makes sense to put it in the `tests/sanity/` directory.

For the same reason, naming the file `ignore.txt` (especially in the directory used by `ansible-test`) could be confusing to users, since the name is so close to the one used by `ansible-test`. There's also the potential for further confusion since the file formats are different.

File: docs/configuring.md [hunk 2]
```diff
@@ -38,6 +38,18 @@ counterparts:
 --8<-- ".ansible-lint"
 ```
 
+## Ignoring rules for entire files
+
+Ansible-lint will load skip rules from `tests/sanity/ignore.txt` file, if one is
+detected, the same as
+[ansible-test](https://docs.ansible.com/ansible/latest/dev_guide/testing/sanity/ignores.html).
```
Comment (mattclay) [inline docs/configuring.md:45]: Since this feature is unrelated to `ansible-test`, I don't think it makes sense to reference it here. Not only is the behavior different, but as mentioned above, the file format is different also.

**Comment by mattclay** [comment]
@ssbarnea Can you clarify what you mean by this part of the summary?

> This will also make it easier to integrate results
> from `ansible-test sanity` directly into ansible-lint, as we will store the ignores into a single location.

**Review by mattclay — CHANGES_REQUESTED** [review]

**Comment by cidrblock** [comment]
Related: https://github.com/ansible/ansible-lint/issues/3012

File: .ansible-lint (hunk 1 above)
Comment (mattclay) [reply .ansible-lint:48]: The references to `tests/sanity/ignore.txt` and `ansible-test` should be removed.

**Review by mattclay — CHANGES_REQUESTED** [review]

File: src/ansiblelint/loaders.py [hunk 3]
```diff
@@ -24,4 +29,29 @@ def yaml_from_file(filepath: str | Path) -> Any:
...
+    """Return a list of rules to ignore."""
+    result = defaultdict(set)
+    if os.path.isfile(filepath):
+        with open(str(filepath), encoding="utf-8") as content:
+            _logger.debug("Loading ignores from %s", filepath)
+            for line in content:
+                entry = line.split("#")[0].rstrip()
+                if entry:
+                    try:
+                        path, rule = entry.split()
```
Comment (cidrblock) [inline src/ansiblelint/loaders.py:42]: if path.endwith(':'), path = path[0:-1] ?

**Review by cidrblock — APPROVED** [review]

**Review by mattclay — COMMENTED** [review]
Now that ansible-test references have been removed, I'll leave the reviewing to the regular ansible-lint contributors.

File: src/ansiblelint/loaders.py (hunk 3 above)
Comment (ssbarnea) [reply src/ansiblelint/loaders.py:42]: Nope, only spaces supported for now. No plan to add unneeded complexity.
